### PR TITLE
[tune] Implement BOHB

### LINKED ISSUE #5100: [tune] Implement BOHB
This is actually done already but requires some reversions in HyperBand (forcing it to have strict barriers).

### PR DESCRIPTION
**Pull request by lisadunlap**
## What do these changes do?
Added TuneBOHB (search), HyperBandForBOHB (early stopping), and an example. 

## Related issue number
Closes #5100 

## Linter

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.

### CODE REVIEW

File: python/ray/tune/examples/bohb_example.py [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+import random
+
+import numpy as np
+import ConfigSpace
+import ConfigSpace.hyperparameters as CSH
+import ConfigSpace.util
+
+import ray
+from ray.tune import Trainable, run, sample_from
+from ray.tune.schedulers import TuneBOHB, HyperBandForBOHB
```
Comment (richardliaw) [inline python/ray/tune/examples/bohb_example.py]: TuneBOHB should go to `ray.tune.suggest`

File: python/ray/tune/examples/bohb_example.py [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+        max_t=100)
+
+    run(MyTrainableClass,
+        name="bohb_test",
+        scheduler=bohb,
+        search_alg=TuneBOHB(
+            config_space,
+            max_concurrent=4,
+            # num_concurrent=100,
+            reward_attr="episode_reward_mean",
```
Comment (richardliaw) [inline python/ray/tune/examples/bohb_example.py]: reward_attr => metric

File: python/ray/tune/schedulers/bohb.py [hunk 3]
```diff
@@ -0,0 +1,62 @@
+import ray
```
Comment (richardliaw) [inline python/ray/tune/schedulers/bohb.py]: can you move this file to `tune/suggest`?

and add the three `__future__` imports to the top?

File: python/ray/tune/schedulers/bohb.py [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+                 max_concurrent=8,
+                 reward_attr="neg_mean_loss",
+                 bohb_config=None):
+        self._max_concurrent = max_concurrent
+        self.trial_to_params = {}
+        self.running = set()
+        self.paused = set()
+        self.reward_attr = reward_attr
+        bohb_config = bohb_config or {}
+        self.bohber = BOHB(space, **bohb_config)
```
Comment (richardliaw) [inline python/ray/tune/schedulers/bohb.py]: oh, we need to import BOHB from HpBandSter

**Comment by 20zinnm** [comment]
How much more needs to be done before this can be merged? Seems almost complete and it's a major feature to be missing. I'm willing to help if needed. @lisadunlap

File: doc/source/tune-searchalg.rst [hunk 5]
```diff
@@ -181,6 +182,30 @@ An example of this can be found in `ax_example.py <https://github.com/ray-projec
...
+.. code-block:: bash
+
+    $ pip install hpbandster
+    $ pip install ConfigSpace
+
+This algorithm requires using the `ConfigSpace search space specification <https://automl.github.io/HpBandSter/build/html/quickstart.html#searchspace>`__. You can use TuneBOHB like follows:
+
+.. code-block:: python
+
+    tune.run(... , search_alg=TuneBOHB(config_space, ... ))
```
Comment (richardliaw) [inline doc/source/tune-searchalg.rst]: you also need HyperBand right?

File: python/ray/tune/suggest/bohb.py [hunk 6]
```diff
@@ -0,0 +1,451 @@
...
+import ConfigSpace.util
+import numpy as np
+import scipy.stats as sps
+import statsmodels.api as sm
+
+from hpbandster.core.base_config_generator import base_config_generator
+from ray.tune.suggest import SuggestionAlgorithm
+
+
+class BOHB(base_config_generator):
```
Comment (richardliaw) [inline python/ray/tune/suggest/bohb.py]: ah sorry, I meant  “importing BOHB from hpbandster” is ideal

File: doc/source/tune-searchalg.rst [hunk 7]
```diff
@@ -181,6 +182,30 @@ An example of this can be found in `ax_example.py <https://github.com/ray-projec
     :show-inheritance:
     :noindex:
 
+BOHB (Bayesian optimization and Hyperband)
+---------------------------------------------------
```
Comment (richardliaw) [inline doc/source/tune-searchalg.rst]: should align with the above string

File: python/ray/tune/suggest/bohb.py [hunk 8]
```diff
@@ -0,0 +1,77 @@
+"""BOHB (Bayesian Optimization with HyperBand)"""
+
+from __future__ import absolute_import
+from __future__ import division
+from __future__ import print_function
+
+from hpbandster import BOHB
```
Comment (richardliaw) [inline python/ray/tune/suggest/bohb.py]: can you do a soft import here? like in `suggest/ax.py`; and correspondingly raise an error in __init__ if not found?

File: python/ray/tune/suggest/bohb.py [hunk 9]
```diff
@@ -0,0 +1,77 @@
...
+            self.running.add(trial_id)
+            return config
+        return None
+
+    def on_trial_result(self, trial_id, result):
+        if trial_id not in self.paused:
+            self.running.add(trial_id)
+        if "budget" in result.get("hyperband_info", {}):
+            hbs_wrapper = self.to_wrapper(trial_id, result)
+            print("adding new result", vars(hbs_wrapper))
```
Comment (richardliaw) [inline python/ray/tune/suggest/bohb.py]: can you remove this?

File: python/ray/tune/suggest/bohb.py [hunk 10]
```diff
@@ -0,0 +1,77 @@
...
+                          result=None,
+                          error=False,
+                          early_terminated=False):
+        del self.trial_to_params[trial_id]
+        if trial_id in self.paused:
+            self.paused.remove(trial_id)
+        elif trial_id in self.running:
+            self.running.remove(trial_id)
+        else:
+            import ipdb; ipdb.set_trace()
```
Comment (richardliaw) [inline python/ray/tune/suggest/bohb.py]: can you remove this?

File: python/ray/tune/suggest/bohb.py [hunk 11]
```diff
@@ -0,0 +1,77 @@
...
+        self.kwargs = {"budget": budget, "config": config.copy()}
+        self.exception = None
+
+class TuneBOHB(SuggestionAlgorithm):
+    """Tune Suggestion Algorithm for BOHB code"""
+
+    def __init__(self,
+                 space,
+                 max_concurrent=8,
+                 metric="neg_mean_loss",
```
Comment (richardliaw) [inline python/ray/tune/suggest/bohb.py]: can you also add the `mode` parameter (which says, "max" or "min") and adjust the result accordingly? You can keep `JobWrapper.result` as `loss`, but the input to that should be negated if "mode == max".

File: python/ray/tune/schedulers/hb_bohb.py [hunk 12]
```diff
@@ -0,0 +1,216 @@
...
+from ray.tune.schedulers.hyperband import HyperBandScheduler, Bracket
+from ray.tune.trial import Trial
+
+logger = logging.getLogger(__name__)
+
+
+class HyperBandForBOHB(HyperBandScheduler):
+    """Implements the HyperBand early stopping algorithm.
+
+    @!!!!!!!!!!!!!! Key changes:
```
Comment (richardliaw) [inline python/ray/tune/schedulers/hb_bohb.py]:
```suggestion
    Key changes:
```

File: python/ray/tune/schedulers/hb_bohb.py [hunk 13]
```diff
@@ -0,0 +1,216 @@
...
+
+from ray.tune.schedulers.trial_scheduler import FIFOScheduler, TrialScheduler
+from ray.tune.schedulers.hyperband import HyperBandScheduler, Bracket
+from ray.tune.trial import Trial
+
+logger = logging.getLogger(__name__)
+
+
+class HyperBandForBOHB(HyperBandScheduler):
+    """Implements the HyperBand early stopping algorithm.
```
Comment (richardliaw) [inline python/ray/tune/schedulers/hb_bohb.py]:
```suggestion
    """Extends HyperBand early stopping algorithm for BOHB.
```

File: python/ray/tune/schedulers/hb_bohb.py [hunk 14]
```diff
@@ -0,0 +1,216 @@
...
+
+        #### MAIN CHANGE HERE!
+        statuses = [(t, t.status) for t in bracket._live_trials]
+        if not bracket.filled() or any(
+                status != Trial.PAUSED for t, status in statuses
+                if t is not trial):
+            print(Counter([status for _, status in statuses]))
+            trial_runner._search_alg.on_pause(trial.trial_id)
+            return TrialScheduler.PAUSE
+        # import ipdb; ipdb.set_trace()
```
Comment (richardliaw) [inline python/ray/tune/schedulers/hb_bohb.py]: remove?

File: python/ray/tune/schedulers/hb_bohb.py [hunk 15]
```diff
@@ -0,0 +1,216 @@
...
+            return TrialScheduler.CONTINUE
+
+        result["hyperband_info"]["budget"] = bracket._cumul_r
+
+        #### MAIN CHANGE HERE!
+        statuses = [(t, t.status) for t in bracket._live_trials]
+        if not bracket.filled() or any(
+                status != Trial.PAUSED for t, status in statuses
+                if t is not trial):
+            print(Counter([status for _, status in statuses]))
```
Comment (richardliaw) [inline python/ray/tune/schedulers/hb_bohb.py]: remove?

File: python/ray/tune/schedulers/hb_bohb.py [hunk 16]
```diff
@@ -0,0 +1,216 @@
...
+        and during each halving phase, bad trials will be stopped while good
+        trials will return to "PENDING"."""
+
+        action = TrialScheduler.PAUSE
+        if bracket.cur_iter_done():
+            if bracket.finished():
+                bracket.cleanup_full(trial_runner)
+                return TrialScheduler.STOP
+
+            good, bad = bracket.successive_halving(self._reward_attr)
```
Comment (richardliaw) [inline python/ray/tune/schedulers/hb_bohb.py]: there's no more `_reward_attr`, so this should be `_metric`. `HyperBand` also takes in a `mode` from its superclass, so I think we just need to do the negation here or something.

**Review by richardliaw — COMMENTED** [review]
Nice! Two last things (in addition to above comments):

Can you add `bohb_example` as a line in `ray/ci/jenkins_tests/run_tune_tests.sh` with `--smoke-test`?

Can you run `scripts/format.sh`? You'll need flake8 and yapf==0.23.0 and flake8-quotes. This will run formatting.

File: python/ray/tune/suggest/bohb.py (hunk 11 above)
Comment (lisadunlap) [reply python/ray/tune/suggest/bohb.py]: Are you sure the input should be negated if the mode is max? It looks like when it is used in other suggestion algorithms it is negated when mode is min.

File: ci/jenkins_tests/run_tune_tests.sh [hunk 17]
```diff
@@ -103,3 +103,7 @@ $SUPPRESS_OUTPUT docker run --rm --shm-size=${SHM_SIZE} --memory=${MEMORY_SIZE}
 $SUPPRESS_OUTPUT docker run --rm --shm-size=${SHM_SIZE} --memory=${MEMORY_SIZE} $DOCKER_SHA \
     python /ray/python/ray/tune/examples/skopt_example.py \
     --smoke-test
+
+$SUPPRESS_OUTPUT docker run --rm --shm-size=${SHM_SIZE} --memory=${MEMORY_SIZE} $DOCKER_SHA \
+    python /ray/python/ray/tune/examples/bohb_example.py \
+    --smoke-test
```
Comment (richardliaw) [inline ci/jenkins_tests/run_tune_tests.sh]: oh you'll also have to install hpbandster and configspace in `docker/examples/Dockerfile`

File: doc/source/tune-searchalg.rst [hunk 18]
```diff
@@ -181,6 +182,30 @@ An example of this can be found in `ax_example.py <https://github.com/ray-projec
     :show-inheritance:
     :noindex:
 
+BOHB (Bayesian optimization and Hyperband)
```
Comment (richardliaw) [inline doc/source/tune-searchalg.rst]: Can you just change this to be BOHB?

File: doc/source/tune-searchalg.rst [hunk 19]
```diff
@@ -181,6 +182,30 @@ An example of this can be found in `ax_example.py <https://github.com/ray-projec
...
+------------------------------------------
+
+The ``BOHB`` is a SearchAlgorithm that is backed by `HpBandSter <https://github.com/automl/HpBandSter>`__ to perform sequential model-based hyperparameter optimization in conjunction with Hyperband. Note that this class does not extend ``ray.tune.suggest.BasicVariantGenerator``, so you will not be able to use Tune's default variant generation/search space declaration when using BOHB. Also note that BOHB is intended to be paried with the Hyperband scheduler `hp_bohb.py <https://github.com/ray-project/ray/blob/master/python/ray/tune/schedulers/hp_bohb.py>`__.
```
Comment (richardliaw) [inline doc/source/tune-searchalg.rst]:
```suggestion
``BOHB`` (Bayesian Optimization HyperBand) is a SearchAlgorithm that is backed by `HpBandSter <https://github.com/automl/HpBandSter>`__ to perform sequential model-based hyperparameter optimization in conjunction with HyperBand. Note that this class does not extend ``ray.tune.suggest.BasicVariantGenerator``, so you will not be able to use Tune's default variant generation/search space declaration when using BOHB. Also note that BOHB is intended to be paired with a specific Hyperband scheduler `ray.tune.schedulers.HyperBandForBOHB` <https://github.com/ray-project/ray/blob/master/python/ray/tune/schedulers/hp_bohb.py>`__.
```

**Comment by richardliaw** [comment]
(Some tests failing)

**Review by richardliaw — APPROVED** [review]

**Comment by richardliaw** [comment]
@20zinnm thanks for offering! This is now merged. Could you try it out and let us know how it goes?

**Pull request merged**